File: data structure/Stack/9012.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int num = scanner.nextInt();

        for (int i = 0; i < num; i++) {
            String str = scanner.next();
            Stack stack = new Stack<>();
            String result = "YES";

            for (int j = 0; j < str.length(); j++) {

                if (str.charAt(j) == '(') {
                    stack.push(str.charAt(j));
                }

                if (str.charAt(j) == ')') {

                    if (!stack.isEmpty()) {
                        stack.pop();
                    } else {
                        result = "NO";
                    }
                }

            }

            if (!stack.isEmpty()) {
                result = "NO";
            }

            System.out.println(result);
        }
    }
}

File: baekjoon/data-structure/9012.py
n = int(input())

for _ in range(n):
    data = list(input())
    stack = []

    while data:
        x = data.pop(0)

        if x == ')' and stack and stack[-1] == '(':
            stack.pop()
        else:
            stack.append(x)
        

    if stack:
        print('NO')
    else:
        print('YES')

######################################################################

for _ in range(n):
    data = list(input())
    sum_ = 0

    for s in data:
        if s == '(':
            sum_ += 1
        else:
            sum_ -= 1

        if sum_ < 0:
            print('NO')
            break

    if sum_:
        print('NO')
    else:
        print('YES')
